### ISSUE TITLE
Extra features for the numeric constants

### ISSUE THREAD

**Issue by tailhook** [edgedb/edgedb#920]
Tho things quite common in modern languages:
1. `123.` for floats (i.e. no digits in decimal part)
2. `1_000_000` for long numbers (underscores)

I'm going to implement it in rust tokenizer since the start if no one objects.

**Comment by 1st1** [comment]
> I'm going to implement it in rust tokenizer since the start if no one objects.

Let's consider this **after** alpha 2, please.  We're all for implementing `1_000_000` -- this was on our mental to-do list.  As for `123.` -- need to check if we have no potential issues with it in our grammar.

**Comment by vpetrovykh** [comment]
> 1. `123.` for floats (i.e. no digits in decimal part)

We have the following allowed expression for accessing nested tuple values:
```
db> WITH a := (1,(2,3)) # could be any nested tuple
... SELECT a.1.1;
{3}
```
The problem is that `1.1` is tokenized by the lexer as a float literal and then there's a parser hack that correctly interprets it as a path instead. This hack (see [here](https://github.com/edgedb/edgedb/blob/eef5085bdbb20e286cc1ee4a0d38c37b54e07b8f/edb/edgeql/parser/grammar/expressions.py#L1009)) depends on knowing that the only valid float for tuple access is of the form `123.456`. Allowing a trailing `.` would complicate the grammar significantly. This begs the question whether the convenience of not typing a `0` is worth the extra complexity. Also, if we allow `1.` should we also allow `.1` as a valid number? Again, this will increase the parser complexity significantly as well.

**Comment by tailhook** [comment]
> The problem is that 1.1 is tokenized by the lexer as a float literal and then there's a parser hack that correctly interprets it as a path instead. 

Well, how common is such access to nested tuples? Rust have the same problem and they decided to forbid it:
```
2 |     println!("{}", (1,).0.2.3)
  |                    -----^^^
  |                    |    |
  |                    |    unexpected token
  |                    help: try parenthesizing the first index: `((1,).0).2`
```
While I'm okay with forbidding `0.` for now. I don't think this parser hack is a good idea in the long term.

**Comment by 1st1** [comment]
Why? It works and makes tuple access syntax obvious for users. On the contrary, `1.` isn't a common way of specifying floats, I rarely see when I'm browsing code in any language.

**Comment by tailhook** [comment]
Another thought: we can always parse `<dot><digits>` as `<dot><integer><dot><integer>` and never as float. This introduces a little bit of state to tokenizer, but the state in tokenizer is quite common pattern.

And this doesn't prohibit `0.`

> Let's consider this after alpha 2, please.

Sure, rust parser isn't going to be integrated before we cut alpha 2.

**Comment by tailhook** [comment]
> Another thought: we can always parse <dot><digits> as <dot><integer><dot><integer>

Like [this](https://github.com/edgedb/edgedb-rust/commit/ee8c555423c31e813fa0d2035f8dc4cabe483cbf) (that last error message could be clearer, though)

**Comment by 1st1** [comment]
> Like this (that last error message could be clearer, though)

Yeah, sure. It should work if we can add a few relatively simple hacks to the lexer to make both `0.` and `.0` **and** tuple paths without parens!

**Comment by vpetrovykh** [comment]
You're forgetting `1.e123` (perfectly fine in JS and Python). What this all means is that there would be several distinct lexer tokens to account for different options and then also messy parser rules (because half of them would have to account for rules terminating in a `DOT`, while the other half would have a `DOT` in the middle). Oh, and then there's also the duplication of some of these rules for `decimal`, like `1.n` and `1.e123n`. The `.123` token would also have to be accounted for in several more rules, the ones dealing with `<.` and `>.`. 

In short, the changes are not as trivial and obvious as they seem and In the end effect for the user is minimal. They would introduce backwards-compatible syntax, if they are needed later.

**Comment by tailhook** [comment]
Current rust tokenizer ([added a test to make sure](https://github.com/edgedb/edgedb-rust/commit/98c32940aabfab548b9a4b77433b1898a1c18f80)) parses as follows:
1. `tuple.1.<` parses fine as `tuple . 1  .<`
2. `1.<` parses as `1. <` but `.<` doesn't make sense on a number as far as I understand
3. similarly with `1.e123` is just a float, but...
4. `xxx.1.e123` is `xxx . 1 . e123` (i.e. tuple access, attribute access)

I'm not sure if (2) might cause confusion for somebody. Fortunately, we don't need to decide it until alpha 2 :)

**Comment by vpetrovykh** [comment]
> `.<` doesn't make sense on a number as far as I understand

You're right, I overhastily included that as valid, but `.>` is supposed to be perfectly equivalent to `.` and valid anywhere where `.` navigation/dereferencing is valid.

**Comment by 1st1** [comment]
@tailhook @vpetrovykh Anything left to do in this one?

**Comment by tailhook** [comment]
Yes #1416
